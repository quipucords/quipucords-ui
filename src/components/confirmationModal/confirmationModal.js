import React from 'react';
import PropTypes from 'prop-types';
import {
  Alert,
  AlertVariant as ConfirmationVariant,
  Button,
  ButtonVariant,
  ModalVariant,
  Title
} from '@patternfly/react-core';
import { Modal } from '../modal/modal';
import { connect, store, reduxTypes } from '../../redux';
import { translate } from '../i18n/i18n';

/**
 * Display a confirmation modal with actions.
 *
 * @param {object} props
 * @param {React.ReactNode} props.body
 * @param {React.ReactNode} props.children
 * @param {string} props.cancelButtonText
 * @param {string} props.confirmButtonText
 * @param {React.ReactNode} props.heading
 * @param {string} props.icon
 * @param {boolean} props.isClose
 * @param {boolean} props.isContentOnly
 * @param {boolean} props.isActions
 * @param {Function} props.onCancel
 * @param {Function} props.onConfirm
 * @param {boolean} props.show
 * @param {Function} props.t
 * @param {string} props.title
 * @param {string} props.variant
 * @returns {React.ReactNode}
 */
const ConfirmationModal = ({
  body,
  children,
  cancelButtonText,
  confirmButtonText,
  heading,
  icon,
  isClose,
  isContentOnly,
  isActions,
  onCancel,
  onConfirm,
  show,
  t,
  title,
  variant
}) => {
  const cancel = () => {
    if (onCancel) {
      onCancel();
    } else {
      store.dispatch({
        type: reduxTypes.confirmationModal.CONFIRMATION_MODAL_HIDE
      });
    }
  };

  const setActions = () => {
    const actions = [];

    if (!isActions) {
      return actions;
    }

    if (onConfirm) {
      actions.push(
        <Button key="submit" onClick={onConfirm}>
          {confirmButtonText || t('form-dialog.label', { context: ['submit', 'confirmation'] })}
        </Button>
      );
    }

    actions.push(
      <Button key="cancel" variant={ButtonVariant.secondary} onClick={cancel}>
        {cancelButtonText || t('form-dialog.label', { context: 'cancel' })}
      </Button>
    );

    return actions;
  };

  const updatedChildren =
    body || heading ? (
      <Alert isInline isPlain variant={icon || undefined} title={heading}>
        <p>{body}</p>
      </Alert>
    ) : (
      children
    );

  return (
    <Modal
      actions={setActions()}
      backdrop={false}
      className="quipucords-modal__confirmation"
      disableFocusTrap
      header={
        title && (
          <Title headingLevel="h4">{title || t('form-dialog.label', { context: ['submit', 'confirmation'] })}</Title>
        )
      }
      isContentOnly={isContentOnly}
      isOpen={show}
      onClose={cancel}
      showClose={isClose}
      variant={variant}
    >
      {updatedChildren || ''}
    </Modal>
  );
};

/**
 * Prop types
 *
 * @type {{isClose: boolean, isActions: boolean, heading: React.ReactNode, icon: string, body: React.ReactNode, title: string,
 *     show: boolean, t: Function, children: React.ReactNode, onCancel: Function, onConfirm: Function, variant: string,
 *     isContentOnly: boolean, cancelButtonText: string, confirmButtonText: string}}
 */
ConfirmationModal.propTypes = {
  body: PropTypes.node,
  cancelButtonText: PropTypes.string,
  children: PropTypes.node,
  confirmButtonText: PropTypes.string,
  heading: PropTypes.node,
  icon: PropTypes.oneOf([...Object.values(ConfirmationVariant)]),
  isActions: PropTypes.bool,
  isClose: PropTypes.bool,
  isContentOnly: PropTypes.bool,
  show: PropTypes.bool.isRequired,
  onCancel: PropTypes.func,
  onConfirm: PropTypes.func,
  t: PropTypes.func,
  title: PropTypes.string,
  variant: PropTypes.oneOf([...Object.values(ModalVariant)])
};

/**
 * Default props.
 *
 * @type {{isClose: boolean, isActions: boolean, heading: null, icon: ConfirmationVariant.warning, title: null, body: null,
 *     t: translate, children: null, onCancel: null, onConfirm: null, variant: null, isContentOnly: boolean,
 *     confirmButtonText: null, cancelButtonText: null}}
 */
ConfirmationModal.defaultProps = {
  children: null,
  title: null,
  heading: null,
  body: null,
  icon: ConfirmationVariant.warning,
  isActions: true,
  isClose: true,
  isContentOnly: false,
  confirmButtonText: null,
  cancelButtonText: null,
  onConfirm: null,
  onCancel: null,
  t: translate,
  variant: null
};

const mapStateToProps = state => ({ ...state.confirmationModal });

const ConnectedConfirmationModal = connect(mapStateToProps)(ConfirmationModal);

export { ConnectedConfirmationModal as default, ConnectedConfirmationModal, ConfirmationModal, ConfirmationVariant };
