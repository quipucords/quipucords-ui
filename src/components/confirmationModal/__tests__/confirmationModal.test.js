import React from 'react';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import { ConnectedConfirmationModal, ConfirmationModal } from '../confirmationModal';

describe('Confirmation Modal Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component', () => {
    const store = generateEmptyStore({
      confirmationModal: {
        show: true,
        title: 'Confirm',
        heading: 'test',
        icon: null,
        body: 'Test body',
        confirmButtonText: 'Confirm',
        cancelButtonText: 'Cancel'
      }
    });
    const component = shallow(
      <Provider store={store}>
        <ConnectedConfirmationModal />
      </Provider>
    );

    expect(component.find(ConnectedConfirmationModal)).toMatchSnapshot('connected');
  });

  it('should display a confirmation modal', async () => {
    const onCancel = jest.fn();
    const props = {
      show: true,
      title: 'Confirm',
      heading: 'test',
      icon: null,
      body: 'Test body',
      confirmButtonText: 'Confirm',
      cancelButtonText: 'Cancel',
      onCancel
    };

    const component = await mountHookComponent(<ConfirmationModal {...props} />);
    expect(component.render()).toMatchSnapshot('show');

    component.find('button[className="pf-c-button pf-m-secondary"]').simulate('click');
    expect(onCancel).toHaveBeenCalledTimes(1);
  });

  it('should NOT display a confirmation modal', async () => {
    const props = {
      show: false
    };

    const component = await mountHookComponent(<ConfirmationModal {...props} />);
    expect(component.render()).toMatchSnapshot('hidden');
  });

  it('should allow passed children, or specific props', async () => {
    const props = {
      show: true,
      heading: 'Lorem ipsum',
      children: 'hello world'
    };

    const component = await mountHookComponent(<ConfirmationModal {...props} />);
    expect(component.find('.pf-c-modal-box__body').render()).toMatchSnapshot('heading');

    component.setProps({
      heading: null,
      body: 'Dolor sit'
    });

    expect(component.find('.pf-c-modal-box__body').render()).toMatchSnapshot('body');

    component.setProps({
      body: null
    });

    expect(component.find('.pf-c-modal-box__body').render()).toMatchSnapshot('children');
  });

  it('should allow custom content', async () => {
    const props = {
      show: true,
      isActions: false,
      isClose: false,
      isContentOnly: true
    };

    const component = await mountHookComponent(<ConfirmationModal {...props}>lorem ipsum</ConfirmationModal>);
    expect(component.render()).toMatchSnapshot('custom');
  });
});
