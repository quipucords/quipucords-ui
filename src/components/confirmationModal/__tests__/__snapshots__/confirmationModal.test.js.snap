// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirmation Modal Component should NOT display a confirmation modal: hidden 1`] = `null`;

exports[`Confirmation Modal Component should allow custom content: custom 1`] = `
<div
  class="pf-c-backdrop"
>
  <div
    class="pf-l-bullseye"
  >
    <div
      aria-label="t(modal.aria-label-default)"
      aria-labelledby="pf-modal-part-3"
      aria-modal="true"
      class="pf-c-modal-box pf-m-align-top"
      data-ouia-component-id="OUIA-Generated-Modal-4"
      data-ouia-component-type="PF4/ModalContent"
      data-ouia-safe="true"
      id="pf-modal-part-3"
      role="dialog"
      style="--pf-c-modal-box--m-align-top--spacer: 5%;"
    >
      <div>
        lorem ipsum
      </div>
    </div>
  </div>
</div>
`;

exports[`Confirmation Modal Component should allow passed children, or specific props: body 1`] = `
<div
  class="pf-c-modal-box__body"
  id="pf-modal-part-4"
>
  <div
    aria-label="Warning Alert"
    class="pf-c-alert pf-m-inline pf-m-plain pf-m-warning"
    data-ouia-component-id="OUIA-Generated-Alert-warning-1"
    data-ouia-component-type="PF4/Alert"
    data-ouia-safe="true"
  >
    <div
      class="pf-c-alert__icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 576 512"
        width="1em"
      >
        <path
          d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        />
      </svg>
    </div>
    <h4
      class="pf-c-alert__title"
    >
      <span
        class="pf-u-screen-reader"
      >
        Warning alert:
      </span>
    </h4>
    <div
      class="pf-c-alert__description"
    >
      <p>
        Dolor sit
      </p>
    </div>
  </div>
</div>
`;

exports[`Confirmation Modal Component should allow passed children, or specific props: children 1`] = `
<div
  class="pf-c-modal-box__body"
  id="pf-modal-part-4"
>
  <div>
    hello world
  </div>
</div>
`;

exports[`Confirmation Modal Component should allow passed children, or specific props: heading 1`] = `
<div
  class="pf-c-modal-box__body"
  id="pf-modal-part-4"
>
  <div
    aria-label="Warning Alert"
    class="pf-c-alert pf-m-inline pf-m-plain pf-m-warning"
    data-ouia-component-id="OUIA-Generated-Alert-warning-1"
    data-ouia-component-type="PF4/Alert"
    data-ouia-safe="true"
  >
    <div
      class="pf-c-alert__icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 576 512"
        width="1em"
      >
        <path
          d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        />
      </svg>
    </div>
    <h4
      class="pf-c-alert__title"
    >
      <span
        class="pf-u-screen-reader"
      >
        Warning alert:
      </span>
      Lorem ipsum
    </h4>
    <div
      class="pf-c-alert__description"
    >
      <p />
    </div>
  </div>
</div>
`;

exports[`Confirmation Modal Component should display a confirmation modal: show 1`] = `
<div
  class="pf-c-backdrop"
>
  <div
    class="pf-l-bullseye"
  >
    <div
      aria-describedby="pf-modal-part-2"
      aria-label="t(modal.aria-label-default)"
      aria-labelledby="pf-modal-part-0"
      aria-modal="true"
      class="pf-c-modal-box pf-m-align-top"
      data-ouia-component-id="OUIA-Generated-Modal-1"
      data-ouia-component-type="PF4/ModalContent"
      data-ouia-safe="true"
      id="pf-modal-part-0"
      role="dialog"
      style="--pf-c-modal-box--m-align-top--spacer: 5%;"
    >
      <button
        aria-disabled="false"
        aria-label="Close"
        class="pf-c-button pf-m-plain"
        data-ouia-component-id="OUIA-Generated-Modal-1-ModalBoxCloseButton"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 352 512"
          width="1em"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          />
        </svg>
      </button>
      <header
        class="pf-c-modal-box__header"
      >
        <h4
          class="pf-c-title pf-m-md"
          data-ouia-component-id="OUIA-Generated-Title-1"
          data-ouia-component-type="PF4/Title"
          data-ouia-safe="true"
        >
          Confirm
        </h4>
      </header>
      <div
        class="pf-c-modal-box__body"
        id="pf-modal-part-2"
      >
        <div
          aria-label="Default Alert"
          class="pf-c-alert pf-m-inline pf-m-plain"
          data-ouia-component-id="OUIA-Generated-Alert-default-1"
          data-ouia-component-type="PF4/Alert"
          data-ouia-safe="true"
        >
          <div
            class="pf-c-alert__icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 896 1024"
              width="1em"
            >
              <path
                d="M448,0 C465.333333,0 480.333333,6.33333333 493,19 C505.666667,31.6666667 512,46.6666667 512,64 L512,106 L514.23,106.45 C587.89,121.39 648.48,157.24 696,214 C744,271.333333 768,338.666667 768,416 C768,500 780,568.666667 804,622 C818.666667,652.666667 841.333333,684 872,716 C873.773676,718.829136 875.780658,721.505113 878,724 C890,737.333333 896,752.333333 896,769 C896,785.666667 890,800.333333 878,813 C866,825.666667 850.666667,832 832,832 L63.3,832 C44.9533333,831.84 29.8533333,825.506667 18,813 C6,800.333333 0,785.666667 0,769 C0,752.333333 6,737.333333 18,724 L24,716 L25.06,714.9 C55.1933333,683.28 77.5066667,652.313333 92,622 C116,568.666667 128,500 128,416 C128,338.666667 152,271.333333 200,214 C248,156.666667 309.333333,120.666667 384,106 L384,63.31 C384.166667,46.27 390.5,31.5 403,19 C415.666667,6.33333333 430.666667,0 448,0 Z M576,896 L576,897.08 C575.74,932.6 563.073333,962.573333 538,987 C512.666667,1011.66667 482.666667,1024 448,1024 C413.333333,1024 383.333333,1011.66667 358,987 C332.666667,962.333333 320,932 320,896 L576,896 Z"
              />
            </svg>
          </div>
          <h4
            class="pf-c-alert__title"
          >
            <span
              class="pf-u-screen-reader"
            >
              Default alert:
            </span>
            test
          </h4>
          <div
            class="pf-c-alert__description"
          >
            <p>
              Test body
            </p>
          </div>
        </div>
      </div>
      <footer
        class="pf-c-modal-box__footer"
      >
        <button
          aria-disabled="false"
          class="pf-c-button pf-m-secondary"
          data-ouia-component-id="OUIA-Generated-Button-secondary-1"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe="true"
          type="button"
        >
          Cancel
        </button>
      </footer>
    </div>
  </div>
</div>
`;

exports[`Confirmation Modal Component should render a connected component: connected 1`] = `<Connect(ConfirmationModal) />`;
